keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1001 @@
+other than latest are applied:
+
+- If the constrained pod field has not changed since the pinned version, the policy is applied as
+  originally specified.
+- The privileged profile always means fully unconstrained and is effectively unversioned (specifying
+  a version is allowed but ignored).
+- Specifying a version more recent than the current Kubernetes version is allowed (for rollback &
+  version skew reasons), but is treated as `latest`.
+
+<<[UNRESOLVED]>>
+
+Under the webhook implementation, policy versions are tied to the webhook version, not the cluster
+version. This means that it is recommended for the webhook to updated prior to updating the cluster.
+
+~~if the webhook version is newer than the cluster version, how
+should policy versions past the cluster version be handled? Should `latest` mean the latest version
+supported by the webhook, or the policy version matching the cluster version? _Leaning towards
+supporting all policy versions supported by the webhook, even if they are newer than the cluster
+version._~~
+
+Note that policies are not guaranteed to be backwards compatible, and a newer restricted policy
+could require setting a field that doesn't exist in the current API version.
+
+<<[/UNRESOLVED]>>
+
+- Under an older version X of a policy:
+    - Allow pods that were allowed under policy version X running on cluster version X
+    - Allow pods that set new fields that the policy level has no opinion about (e.g. pod overhead
+      fields)
+    - Allow pods that set new fields that the policy level has an opinion about if the value is the
+      default (explicit or implicit) value OR the value is allowed by newer versions of the policy
+      level (e.g. a less privileged value of a new field)
+
+For example, the restricted policy level now requires `allowPrivilegeEscalation=false`, but this
+field wasn't added until Kubernetes v1.8, and all containers prior to v1.8 implicitly ran as
+`allowPrivilegeEscalation=true`. Under the **restricted v1.7** profile, the following
+`allowPrivilegeEscalation` configurations would be allowed on a v1.8 cluster:
+- `null` (allowed during the v1.7 release)
+- `true` (equal in privilege to a v1.7 pod that didn't set the field)
+- `false` (strictly less privileged than other allowed values)
+
+### PodTemplate Resources
+
+Audit and Warn modes are also checked on resource types that embed a PodTemplate (enumerated below),
+but Allow mode only applies to actual pod resources.
+
+Since users do not create pods directly in the typical deployment model, the warning mechanism is
+only effective if it can also warn on templated pod resources. Similarly, for audit it is useful to
+tie the audited violation back to the requesting user, so audit will also apply to templated pod
+resources. In the interest of supporting mutating admission controllers, policies will only
+be enforced on actual pods.
```

Comment:
Sorry, probably I'm missing the obvious, but can't it be placed after mutating admission controllers, similar to Mutating Webhook and Validating Webhook?

For us part of the pain with PSPs is, that developers lack early feedback: The deployment is applied without problems, but the pods are not coming up. With this proposal they could at least get a warning, but for us it would be nice to directly fail the kubectl apply.

Reply:
Most mutating admission controls don't mutate pod specs in workload types like Deployment, they mutate directly on the created pods. Failing `apply` of a Deployment because the podspec didn't conform to the policy introduces the possibility of false positive denials, where the Deployment is rejected but some subsequent mutating admission plugin would have set fields on the pod that would have allowed it to pass the policy.

Reply:
Usually this shows up in the events of the replicaset. This is a bit confusing to users though. Maybe a solution would be to propagate the events up to deployments too, so a describe deployment would show the pod spec issues too? Then it wouldn't be so bad if just the pod was validated.

Reply:
Warnings will apply to deployments, so the idea would be that you could add a warning for the early feedback, even if it doesn't give you a hard denial. I think events should be propagated up already?

Reply:
I haven't used really recent versions of k8s, so maybe its fixed now. but I've been befuddled before when there isn't enough quota and the events only show up in the replicaset and I forget to look there.

Reply:
> I've been befuddled before when there isn't enough quota and the events only show up in the replicaset and I forget to look there.

that would be good feedback for sig-apps (replicaset/deployment controllers) or sig-cli (describe command), but is orthogonal here